src/queue.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::{HashSet, VecDeque};
+use tracing::{debug, warn};
+
+use crate::message::TransportMessage;
+
+/// MessageQueue is a queue of messages, ordered by nonce, that we've
+/// received but are not yet able to process because we're waiting for
+/// a message with the next expected nonce first.
+/// This is required because Nym does not guarantee any sort of message
+/// ordering, only delivery.
+/// TODO: is there a DOS vector here where a malicious peer sends us
+/// messages only with nonce higher than the next expected nonce?
+pub(crate) struct MessageQueue {
+    /// nonce of the next message we expect to receive on the
+    /// connection.
+    /// any messages with a nonce greater than this are pushed into
+    /// the queue.
+    /// if we get a message with a nonce equal to this, then we
+    /// immediately handle it in the transport and increment the nonce.
+    next_expected_nonce: u64,
+
+    /// the actual queue of messages, ordered by nonce.
+    /// the head of the queue's nonce is always greater
+    /// than the next expected nonce.
+    queue: VecDeque<TransportMessage>,
+
+    /// tracks the nonces that exist in the queue.
+    /// used to check for duplicate nonces on push.
+    nonce_set: HashSet<u64>,
+}
+
+impl MessageQueue {
+    pub(crate) fn new() -> Self {
+        MessageQueue {
+            next_expected_nonce: 0,
+            queue: VecDeque::new(),
+            nonce_set: HashSet::new(),
+        }
+    }
+
+    pub(crate) fn print_nonces(&self) {
+        let mut nonces = "".to_string();
+        self.queue.iter().for_each(|msg| {
+            nonces += &msg.nonce.to_string();
+            nonces += ", ";
+        });
+        debug!("MessageQueue: [{:?}]", nonces);
+    }
+
+    /// sets the next expected nonce to 1, indicating that we've received
+    /// a ConnectionRequest or ConnectionResponse.
+    pub(crate) fn set_connection_message_received(&mut self) {
+        if self.next_expected_nonce != 0 {
+            panic!("connection message received twice");
+        }
+
+        self.next_expected_nonce += 1;
+    }
+
+    /// tries to push a message into the queue.
+    /// if the message has the next expected nonce, then the message is returned,
+    /// and should be processed by the caller.
+    /// in that case, the internal queue's next expected nonce is incremented.
+    pub(crate) fn try_push(&mut self, msg: TransportMessage) -> Option<TransportMessage> {
```

Comment:
There should be a `mock` test here to deliberately send the messages with nonce in out of order.

Reply:
done